test/test_sync.py

```diff
@@ -105,6 +115,82 @@ def test_invalid_permissions(self):
                 os.path.join(tmpdir, self.NAMES[0])
             )
 
+    def _check_file_filters_results(self, filters_manager, expected_scan_results):
+        with TempDir() as tmpdir:
+            folder = self._prepare_folder(tmpdir)
+            self.assertEqual(
+                expected_scan_results,
+                list(f.name for f in folder.all_files(self.reporter, filters_manager))
+            )
+            self.reporter.local_access_error.assert_not_called()
+
+    def test_exclusions(self):
+        expected_list = [
+            six.u('.dot_file'),
+            six.u('hello.'),
+            six.u('hello/a/1'),
+            six.u('hello/a/2'),
+            six.u('hello/b'),
+            six.u('hello0'),
+            six.u('inner/a.txt'),
+            six.u('inner/b.txt'),
+            six.u('inner/more/a.txt'),
+            six.u('\u81ea\u7531'),
+        ]
+        filters_manager = FilterManager([
+            ExcludeFileRegexFilter('.*\\.bin'),
+        ])
+        self._check_file_filters_results(filters_manager, expected_list)
+
+    def test_exclude_all(self):
+        expected_list = []
+        filters_manager = FilterManager([
+            ExcludeFileRegexFilter('.*'),
+        ])
+        self._check_file_filters_results(filters_manager, expected_list)
+
+    def test_exclusions_inclusions(self):
+        expected_list = [
+            six.u('.dot_file'),
+            six.u('hello.'),
+            six.u('hello/a/1'),
+            six.u('hello/a/2'),
+            six.u('hello/b'),
+            six.u('hello0'),
+            six.u('inner/a.bin'),
+            six.u('inner/a.txt'),
+            six.u('inner/b.txt'),
+            six.u('inner/more/a.bin'),
+            six.u('inner/more/a.txt'),
+            six.u('\u81ea\u7531'),
+        ]
+        filters_manager = FilterManager(
+            [
+                ExcludeFileRegexFilter('.*\\.bin'),
+                IncludeFileRegexFilter('.*a\\.bin'),
+            ]
+        )
+        self._check_file_filters_results(filters_manager, expected_list)
+
+    def test_exclude_directory(self):
+        expected_list = [
+            six.u('.dot_file'),
+            six.u('hello.'),
+            six.u('hello0'),
+            six.u('inner/a.bin'),
+            six.u('inner/a.txt'),
+            six.u('inner/b.bin'),
+            six.u('inner/b.txt'),
+            six.u('\u81ea\u7531'),
+        ]
+        filters_manager = FilterManager(
+            [
+                ExcludeDirRegexFilter('.*hello'),
+                ExcludeDirRegexFilter('.*more'),
```

Comment:
How about a test that an excludeFile regex doesn't exclude a directory?  Adding `ExcludeFileRegexFilter('innner')` would test that.

Reply:
And also a test of the inverse would be good.  `ExcludeDirRegexFilter('hello0')` should not exclude the file `hello0`